tests/SimpleSAML/DatabaseTest.php

```diff
@@ -0,0 +1,245 @@
+	 * @covers SimpleSAML_Database::generateInstanceId
+	 * @covers SimpleSAML_Database::__construct
+	 * @covers SimpleSAML_Database::connect
+	 * @covers SimpleSAML_Database::getSlave
+	 * @test
+	 */
+	public function Slaves(){
+		$getSlave = self::getMethod('getSlave');
+		
+		$master = spl_object_hash(PHPUnit_Framework_Assert::readAttribute($this->db, 'dbMaster'));
+		$slave = spl_object_hash($getSlave->invokeArgs($this->db, array()));
+
+		$this->assertTrue(($master == $slave), "getSlave should have returned the master database object");
+
+		$config = array(
+			'database.dsn' => 'sqlite::memory:',
+			'database.username' => null,
+			'database.password' => null,
+			'database.prefix' => 'phpunit_',
+			'database.persistent' => true,
+			'database.slaves' => array(
+				array(
+					'dsn' => 'sqlite::memory:',
+					'username' => null,
+					'password' => null,
+				),
+			),
+		);
+
+		$sspConfiguration = new SimpleSAML_Configuration($config, "test/SimpleSAML/DatabaseTest.php");
+		$msdb = SimpleSAML_Database::getInstance($sspConfiguration);
+
+		$slaves = PHPUnit_Framework_Assert::readAttribute($msdb, 'dbSlaves');
+		$gotSlave = spl_object_hash($getSlave->invokeArgs($msdb, array()));
+
+		$this->assertEquals(spl_object_hash($slaves[0]), $gotSlave, "getSlave should have returned a slave database object");
+	}
+
+	/**
+	 * @covers SimpleSAML_Database::applyPrefix
+	 * @test
+	 */
+	public function prefix(){
+		$prefix = $this->config->getValue('database.prefix');
+		$table = "saml20_idp_hosted";
+		$pftable = $this->db->applyPrefix($table);
+
+		$this->assertEquals($prefix . $table, $pftable, "Did not properly apply the table prefix");
+	}
+
+	/**
+	 * @covers SimpleSAML_Database::write
+	 * @covers SimpleSAML_Database::read
+	 * @covers SimpleSAML_Database::exec
+	 * @covers SimpleSAML_Database::query
+	 * @test
+	 */
+	public function Querying()
+	{
+		$table = $this->db->applyPrefix("sspdbt");
+		$this->assertEquals($this->config->getValue('database.prefix') . "sspdbt", $table);
+
+		$this->db->write("CREATE TABLE IF NOT EXISTS $table (ssp_key INT(16) NOT NULL, ssp_value TEXT NOT NULL)", false);
+
+		$query1 = $this->db->read("SELECT * FROM $table");
+		$this->assertEquals(0, $query1->fetch(), "Table $table is not empty when it should be.");
+
+		$ssp_key = time();
+		$ssp_value = md5(rand(0,10000));
+		$stmt = $this->db->write("INSERT INTO $table (ssp_key, ssp_value) VALUES (:ssp_key, :ssp_value)", array('ssp_key' => array($ssp_key, PDO::PARAM_INT), 'ssp_value' => $ssp_value));
+		$this->assertEquals(1, $stmt->rowCount(), "Could not insert data into $table.");
+
+		$query2 = $this->db->read("SELECT * FROM $table WHERE ssp_key = :ssp_key", array('ssp_key' => $ssp_key));
+		$data = $query2->fetch();
+		$this->assertEquals($data['ssp_value'], $ssp_value, "Inserted data doesn't match what is in the database");
+	}
+
+	/**
+	 * @covers SimpleSAML_Database::read
+	 * @covers SimpleSAML_Database::query
+	 * @expectedException Exception
```

Comment:
I believe that would be a better option rather than using the generic Exception since most of SimpleSAMLphp uses exception handling. Do we want to include adding an extension of the built-in Exception class within this pull request, or handle that in a separate one which could include reworking `SimpleSAML_Error_Exception` to be a bit more flexible so we could have something along the lines of:

* \SimpleSAML\Exception\Database for database errors
* \SimpleSAML\Exception\Metadata for anything within the metadata storage sources/handlers
* \SimpleSAML\Exception\Session for session store issues
* \SimpleSAML\Exception\Generic
* etc.

At this point I'm inclined to say leave it as Exception for now and then create a separate pull request for exception handling as a whole, but I'm up for either or.